src/components/Chat/ChatInterface.tsx

```diff
@@ -184,11 +235,112 @@ function getHistory(msgs: Message[]) {
     for (let i = 0; i < msgs.length; i++) {
         const msg = msgs[i];
         if (msg.type === 'bot') {
-            history += "AI: " + msg.message + "\n";
+            history += "tutor: " + msg.message + "\n";
         } else {
-            history += "User: " + msg.message + "\n";
+            history += "student: " + msg.message + "\n";
         }
     }
 
     return history;
-}
\ No newline at end of file
+}
+
+function renderTutorResponseRealTime(data: ChatResponseStream[]) {
+    const tutorResponse = data[data.length - 1].tutor_response;
+    return tutorResponse;
+}
+
+
+function renderTutorResponseFinal(tutorResponse: string, followUpQuestion: string, questionCompleted: string, levelUndecided: boolean) {
+    if (followUpQuestion !== '' && (levelUndecided || questionCompleted === 'false' && !levelUndecided)) {
+        return tutorResponse + '\n\n' + followUpQuestion;
+    } else {
+        return tutorResponse;
+    }
+}
+
+function getQuestionToShow(questionLevel: string | undefined, rawQuestionData: { [key: string]: string[] } | undefined) {
+    // do not show question when there are no initial questions or no questions at a given level
+    console.log(rawQuestionData, questionLevel);
+    if (!rawQuestionData || !questionLevel || !(questionLevel in rawQuestionData) || rawQuestionData[questionLevel].length === 0) {
+        return undefined;
+    }
+
+    // show the most recent question at a given level
+    const questions = rawQuestionData[questionLevel];
+    return questions[questions.length - 1];
+}
+
+function updateMsgList(
+    msg: string,
+    role: string,
+    msgData: Message[],
+    existingMsgElems: React.JSX.Element[],
+    setMsgs: (msgs: React.JSX.Element[]) => void,
+    setRawMsgs: (msgs: Message[]) => void
+): [React.JSX.Element[], Message[]] {
+
+    const msgElem = role === "AI Tutor" ?
+        (
+            <MessageOther
+                message={msg}
+                timestamp=""
+                displayName="AI Tutor"
+                avatarDisp={true}
+            />
+        ) : (
+            <MessageSelf
+                message={msg}
+                timestamp=""
+                displayName="User"
+                avatarDisp={false}
+            />
+        );
+
+    msgData.push({
+        type: role === "AI Tutor" ? 'bot' : 'user',
+        message: msg
+    });
+
+    const newMsgElems = [
+        msgElem,
+        ...existingMsgElems
+    ];
+
+    setMsgs(
+        newMsgElems
+    );
+
+    setRawMsgs(msgData);
+
+    return [newMsgElems, msgData];
+}
+
+async function getInitQuestionByLevel(taxonomy: string[]) {
```

Comment:
Consider replacing this with a list of predefined questions that are defined as constants.